src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -1260,6 +1260,65 @@ internal virtual FieldSymbol FixedElementField
         /// <returns>True if this is an interface type.</returns>
         internal abstract bool IsInterface { get; }
 
+        /// <summary>
+        /// Verify if the given type can be used to back a tuple type and return cardinality of that tuple type in <paramref name="tupleCardinality"/>. 
+        /// </summary>
+        /// <param name="tupleCardinality">If method returns true, contains cardinality of the compatible tuple type.</param>
+        /// <returns></returns>
+        public override bool IsTupleCompatible(out int tupleCardinality)
+        {
+            if (IsTupleType)
+            {
+                tupleCardinality = TupleElementTypes.Length;
+                return true;
+            }
+
+            // Should this be optimized for perf (caching for VT<0> to VT<7>, etc.)?
+            if (!IsUnboundGenericType &&
+                ContainingSymbol.Kind == SymbolKind.Namespace &&
+                Name == TupleTypeSymbol.TupleTypeName && ContainingNamespace.Name == "System" &&
+                ContainingNamespace.ContainingNamespace?.IsGlobalNamespace == true)
+            {
+                int arity = Arity;
+
+                if (arity > 0 && arity < TupleTypeSymbol.RestPosition)
+                {
+                    tupleCardinality = arity;
+                    return true;
+                }
+                else if (arity == TupleTypeSymbol.RestPosition && !IsDefinition)
+                {
+                    // Skip through "Rest" extensions
+                    TypeSymbol typeToCheck = this;
+                    int levelsOfNesting = 0;
+
+                    do
+                    {
+                        levelsOfNesting++;
+                        typeToCheck = ((NamedTypeSymbol)typeToCheck).TypeArgumentsNoUseSiteDiagnostics[TupleTypeSymbol.RestPosition - 1];
+                    }
+                    while (typeToCheck.OriginalDefinition == this.OriginalDefinition && !typeToCheck.IsDefinition);
+
+                    if (typeToCheck.IsTupleType)
+                    {
+                        tupleCardinality = (TupleTypeSymbol.RestPosition - 1) * levelsOfNesting + typeToCheck.TupleElementTypes.Length;
+                        return true;
+                    }
```

Comment:
This part is getting strengthened in order to ensure the tuple is based on the same VT definition, if it is longer than 7 items long.